src/plopp/widgets/cut3d.py

```diff
@@ -0,0 +1,198 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+
+from .styling import BUTTON_LAYOUT
+from ..core import Node, node
+
+import ipywidgets as ipw
+import numpy as np
+import scipp as sc
+import asyncio
+
+
+class Timer:
+    """
+    From:
+    https://ipywidgets.readthedocs.io/en/8.0.2/examples/Widget%20Events.html#Debouncing
+    """
+
+    def __init__(self, timeout, callback):
+        self._timeout = timeout
+        self._callback = callback
+
+    async def _job(self):
+        await asyncio.sleep(self._timeout)
+        self._callback()
+
+    def start(self):
+        self._task = asyncio.ensure_future(self._job())
+
+    def cancel(self):
+        self._task.cancel()
+
+
+def debounce(wait):
+    """
+    Decorator that will postpone a function's
+    execution until after `wait` seconds
+    have elapsed since the last time it was invoked.
+
+    From:
+    https://ipywidgets.readthedocs.io/en/8.0.2/examples/Widget%20Events.html#Debouncing
+    """
+
+    def decorator(fn):
+        timer = None
+
+        def debounced(*args, **kwargs):
+            nonlocal timer
+
+            def call_it():
+                fn(*args, **kwargs)
+
+            if timer is not None:
+                timer.cancel()
+            timer = Timer(wait, call_it)
+            timer.start()
+
+        return debounced
+
+    return decorator
+
+
+class Cut3dTool(ipw.HBox):
+
+    def __init__(self,
+                 *nodes,
+                 limits,
+                 direction,
+                 view,
+                 value: bool = False,
+                 color='red',
+                 linewidth=1.5,
+                 **kwargs):
+        """
+        """
+        import pythreejs as p3
+        self._limits = limits
+        self._direction = direction
+        axis = 'xyz'.index(self._direction)
+        self._dim = self._limits[axis].dim
+        self._unit = self._limits[axis].unit
+        self._view = view
+
+        w_axis = 2 if self._direction == 'x' else 0
+        h_axis = 2 if self._direction == 'y' else 1
+        width = (self._limits[w_axis][1] - self._limits[w_axis][0]).value
+        height = (self._limits[h_axis][1] - self._limits[h_axis][0]).value
```

Comment:
I don't understand what these parameters are. What do 'x' and 'y' mean? And why is there no 'z'?

Reply:
So in order to make a frame/outline, we have to first make the outline (which by default has the z direction as the vector normal to the plane), and then we rotate it. I have used the following scheme:

x as normal vector:
- width: dz (axis=2)
- height: dy (axis=1)
- rotate around Y axis

y as normal vector:
- width: dx (axis=0)
- height: dz (axis=2)
- rotate around X axis

z as normal vector:
- width: dx (axis=0)
- height: dy (axis=1)
- no rotate